feat: timeouts and manual redirections

quadratz opened the pull request:
- update dependencies
- fix error: maximum limit of redirects
- add timeout and delay

https://github.com/grammyjs/website/pull/961#issuecomment-1851625755

quadratz committed:
update

- update dependencies
- fix error: maximum limit of redirects
- add timeout and delay

dcdunkan commented on fetch.ts:
```diff
@@ -28,18 +35,50 @@ const FETCH_OPTIONS: FetchOptions = {
+      if (retries > 0) {
+        console.log(`Retrying (${retries}) after ${retryDelay / 1000} seconds`);
+        await new Promise((resolve) => setTimeout(resolve, retryDelay));
+        retryDelay = Math.min(retryDelay * 2, maxDelay);
```
I understand that having a small delay is good before retrying. But, I don't see how it is really useful to wait twice the delay of the previous retry. 2, 4, 8, 16, 30. That is just a waste of time I guess. I know currently the link checker doesn't even have timeouts and I agree that it is even worse. But since we're improving that, I don't think we really need this kind of waiting. Please let me know if there is an actual use case, other than hoping that waiting with even longer delays each time, might make the website respond as expected. If that's the case, we can maybe consider increasing the delay a little bit rather than waiting twice longer on each retry.

dcdunkan commented on fetch.ts:
```diff
@@ -28,18 +35,50 @@ const FETCH_OPTIONS: FetchOptions = {
       }
       retries++;
-    } while (retries < maxRetries && response == null);
+    } while (retries <= maxRetries && response == null);
```
```suggestion
    } while (retries < maxRetries && response == null);
```

~~`<=` makes 6 retries in total, ignoring `maxRetries` (= 5), because `retries` starts from `0`~~. My mistake.

dcdunkan commented on fetch.ts:
```diff
@@ -28,18 +35,50 @@ const FETCH_OPTIONS: FetchOptions = {
+      }
       try {
-        response = await fetch(url, options);
+        const controller = new AbortController();
```
I think since our only goal here is to just timeout the request, we can replace `AbortController` with `AbortSignal` and get rid of the `[set,clear]Timeout`s, and make this simpler.

```ts
const timeoutSignal = AbortSignal.timeout(retryDelay);
await fetch(url, { ...options, redirect: "manual", signal: timeoutSignal })
```

dcdunkan commented on fetch.ts:
```diff
@@ -28,18 +35,50 @@ const FETCH_OPTIONS: FetchOptions = {
-        console.log(`Retrying (${retries + 1})`);
+
+        // Check if the error is due to timeout
+        if (err.name === "AbortError") {
```
Need to be changed to `TimeoutError` if we're changing to `AbortSignal.timeout`.

dcdunkan commented on utilities.ts:
```diff
@@ -77,7 +77,7 @@ export async function checkExternalUrl(url: string, utils: { domParser: DOMParse
   }
 
-  if (response.redirected && !isValidRedirection(new URL(transformed), new URL(response.url))) {
+  if (response.isRedirected && !isValidRedirection(new URL(transformed), new URL(response.url))) {
```
There's a catch. You set `isRedirected` to `true` in manual redirections. But in the `response` of manual redirections, the `url` property won't be the redirected URL. It'll be the same URL as we requested, and that is by design.

We could
1. either extend the response by adding one more property, perhaps `redirectedUrl`, and assign the value of `Location` header (if it exists, otherwise the requested URL) in the response. This gives us an opportunity to make sure that it redirects to an expected page even when redirect is set to `"manual"`,
2. or we could just check if the URL exists in `MANUAL_REDIRECTIONS` here as well. If it is, we should ignore.

I'd go with 1 for now, and if gets too complicated to manage these redirects sometime in the future, we can switch to 2.

dcdunkan commented on group_links.ts:
```diff
@@ -12,7 +12,7 @@
-import { blue } from "https://deno.land/std@0.194.0/fmt/colors.ts";
+import { blue } from "https://deno.land/std@0.208.0/fmt/colors.ts";
 import { DOMParser } from "./deps/deno_dom.ts";
 import { yellow } from "./deps/std/fmt.ts";
```
```suggestion
import { DOMParser } from "./deps/deno_dom.ts";
import { blue, yellow } from "./deps/std/fmt.ts";
```

That'd be better. Sorry, I missed this import before.

dcdunkan commented on fetch.ts:
```diff
@@ -85,7 +124,7 @@ export function isValidRedirection(from: URL, to: URL) {
       // Login redirections; e.g., Firebase Console -> Google Account Login
       (
-        (to.hostname === "accounts.google.com" && segments.to[2] === "signin") || // Google
+        (to.hostname === "accounts.google.com" && (segments.to[2] === "signin" || segments.to[1] === "signup")) || // Google
```
We seems to get different paths from that redirect.

> The redirect in my browser is https://accounts.google.com/signup/v2/createaccount?theme=glif&flowName=GlifWebSignIn&flowEntry=SignUp.

> Mine actually is https://accounts.google.com/lifecycle/steps/signup/name with some long query parameters added to that.

Whatever the path is, it seems to have the "signin" or "signup" somewhere, not necessarily at the beginning. So, I think a simple `.includes` check would do.

```suggestion
        (to.hostname === "accounts.google.com" && (segments.to.includes("signin") || segments.to.includes("signup"))) || // Google
```

dcdunkan reviewed (changes requested):
Hi, thank you for spending your time on this. And sorry about the late review. The implementation generally looks good. But I have left some suggestions. Need your opinion as well. Thanks again.

dcdunkan committed:
fix: re-use import

dcdunkan committed:
fix: cover more google auth redirections

dcdunkan committed:
refactor: handling manual redirections

added a small way to access redirectedUrl for manual redirections
incase we need it sometime in the future

dcdunkan committed:
bump deps

dcdunkan committed:
fix: the response info

dcdunkan committed:
unwanted log

dcdunkan committed:
fix: unwanted entry

dcdunkan commented:
@quadratz I have made the changes and will be merging this now. Thanks again for the contribution!

quadratz reviewed (commented):
LGTM. (Sorry for the late reply) >_<
